### PR TITLE
Fix unnecessary switching of map representation for DiscardUnknownFields

### PR DESCRIPTION
**Pull request by gerben-stavenga**
When unknown fields are discarded using reflection in reflection_ops.cc, unnecessary switching of the internal map representation is triggered.

A map<int,int> field cannot have unknown fields. However, in the current code the check used is

if (field->IsMap() && IsMapValueMessageTyped(field)) {
  // Discard map field 
} else {
  // Discard repeated message field
}

This means that the code proceeds discarding unknown fields for each mapentry using the RepeatedPtrField of MapEntry representation of the map. Causing unnecessary work and unnecessary, potentially big, memory allocation. This PR fixes this.

See also issue https://github.com/protocolbuffers/protobuf/issues/13222

This is also a bug fix. When a map<int, MyProto> is in a dirty state, previously it would be skipped. After this change it will correctly use the repeated path to clean the MyProto sub-entries

### CODE REVIEW

**Review by fowles — APPROVED** [review]

**Comment by gerben-stavenga** [comment]
It fails on internal google tests, which is hard for me to debug. It might be tests relying on the current wrong behavior?

**Comment by fowles** [comment]
that isn't a failure, it is waiting for internal review by someone.  The FFI failure looks unrelated.  I would give the internal reviewer a day or two to review it.  No action needed from you unless this goes quiet for a week.

**Pull request closed without merging**
